### PR TITLE
[InstCombine] Fold X * (2^N + 1) >> N -> X + X >> N, or directly to X if X >> N is 0

### PR DESCRIPTION
**Pull request by AZero13**
Alive2 Proofs:
https://alive2.llvm.org/ce/z/eSinJY
https://alive2.llvm.org/ce/z/sweDgc
https://alive2.llvm.org/ce/z/-2dXZi
https://alive2.llvm.org/ce/z/kx2PhF
https://alive2.llvm.org/ce/z/e9QjM6

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (topperc) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1421]: Does this transform require `nuw`? This is nested in a `match(Op0, m_NUWMul`.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (topperc) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1423]: You create an NUW add here, but your proof shows an NSW add

**Comment by topperc** [comment]
The `tgt` function in your Alive2 proof has a dead instruction in it that is distracting to reviewing the proof.

**Comment by AtariDreams** [comment]
> A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> 
> 
> 
> Proof: https://alive2.llvm.org/ce/z/u3y_-F

I addressed this and edited the commit. Please see.

**Comment by AtariDreams** [comment]
Also see: https://alive2.llvm.org/ce/z/NGZx9f

**Comment by topperc** [comment]
> A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.

Proposed where?

**Comment by AtariDreams** [comment]
The ashr is a leftover from a previous PR but they can work as a negative test for now because I am working on another patch that deals with ashr and I might merge it with this one if it is best.

**Comment by AtariDreams** [comment]
@topperc I fixed the proofs.

**Comment by AtariDreams** [comment]
@dtcxzyw Thoughts on this?

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (topperc) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1426]: This preserves the exact flag, but the exact flag is not used in your proofs

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1426]: Now I do.

**Comment by AtariDreams** [comment]
@dtcxzyw You think you could run benchmarks on this?

**Comment by AtariDreams** [comment]
@dtcxzyw would you be so kind as to benchmark this

**Comment by AtariDreams** [comment]
@dtcxzyw would you be so kind as to benchmark this

**Comment by AtariDreams** [comment]
@topperc what do you think about his?

**Comment by topperc** [comment]
@AtariDreams why do you keep force pushing patches? The github UI doesn't make it easy to see if you changed anything in the code in the patch or just rebased.

**Comment by AtariDreams** [comment]
> @AtariDreams why do you keep force pushing patches? The github UI doesn't make it easy to see if you changed anything in the code in the patch or just rebased. If you don't write any comment to explain the what you did, the reviewer is left guessing.

Edited the commit message is what I did.

**Comment by nikic** [comment]
@AtariDreams As you have already been told before: The commit messages do not matter. Only the pull request description will be used. Please refrain from force-pushing to change the commit description.

**Comment by AtariDreams** [comment]
Alright, I just fixed a couple of whitespace issues, and now this is good to go!

**Comment by topperc** [comment]
> Alright, I just fixed a couple of whitespace issues, and now this is good to go!

And yet again you force pushed despite being repeatedly told not to.

**Comment by AtariDreams** [comment]
@dtcxzyw would you be so kind as to benchmark this

**Comment by AtariDreams** [comment]
Added negative tests!

**Comment by AtariDreams** [comment]
I had to add those negative tests to precommit

**Comment by AtariDreams** [comment]
@topperc @dtcxzyw What do you think about this PR?

File: llvm/test/Transforms/InstCombine/ashr-lshr.ll
Comment (topperc) [inline llvm/test/Transforms/InstCombine/ashr-lshr.ll:654]: Drop the space after `@use`

**Comment by topperc** [comment]
> > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> 
> Proposed where?

This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.

**Comment by nikic** [comment]
> > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > 
> > 
> > Proposed where?
> 
> This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.

A variant of this PR was originally part of https://github.com/llvm/llvm-project/pull/88193. What was "proposed" is that these unrelated changes be removed from that PR.

**Comment by AtariDreams** [comment]
> > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> 
> > > 
> 
> > > 
> 
> > > Proposed where?
> 
> > 
> 
> > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> 
> 
> 
> A variant of this PR was originally part of https://github.com/llvm/llvm-project/pull/88193. What was "proposed" is that these unrelated changes be removed from that PR.

And those unrelated changes were!

**Comment by topperc** [comment]
> > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > Proposed where?
> > 
> > 
> > > 
> > 
> > 
> > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > 
> > 
> > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> 
> And those unrelated changes were removed!

Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.

**Comment by AtariDreams** [comment]
> > > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Proposed where?
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > > 
> > > 
> > > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> > 
> > 
> > And those unrelated changes were removed!
> 
> Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.



> > > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Proposed where?
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > > 
> > > 
> > > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> > 
> > 
> > And those unrelated changes were removed!
> 
> Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.



> > > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Proposed where?
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > > 
> > > 
> > > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> > 
> > 
> > And those unrelated changes were removed!
> 
> Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.

Fixed!

**Comment by AtariDreams** [comment]
> > > > > > A generalization of the proposed x * 3/2 -> x + (x >> 1) transformation.
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > Proposed where?
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > This question was not answered. The way this is written implies there was some discussion separate from this PR that proposed something. There should be a link to that proposal or this commit message should be rephrased.
> > > 
> > > 
> > > A variant of this PR was originally part of #88193. What was "proposed" is that these unrelated changes be removed from that PR.
> > 
> > 
> > And those unrelated changes were removed!
> 
> Great. But that doesn't change the fact that the description of this PR needs to be modified. It will become a commit message that will live in the LLVM repo forever. It doesn't look like the fact that this code was previously part of a different PR is relevant to understanding this PR so the sentence should just be deleted.

I removed the sentence from the commit!

**Comment by AtariDreams** [comment]
@topperc I believe this is ready!

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1468]: As this transform introduces multiple uses of X, do we need a freeze here?
I don't know why alive2 doesn't complain about this.

**Review by dtcxzyw — APPROVED** [review]
LGTM if https://github.com/llvm/llvm-project/pull/90295/files#r1602406923 is not an issue. Thank you!

File: llvm/test/Transforms/InstCombine/ashr-lshr.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/ashr-lshr.ll:669]: This is missing tests for numbers other than `3` and `1`. You should test both another combination that works and one that doesn't.

It's also worth noting that this can still be further generalized. You now handle something like `* 3 >> 1`, but this also holds for `* 6 >> 2` or `* 12 >> 3`.

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:669]: For that one, I was in fact thinking about that, but I think that could best be done in another PR because why stop there?

* 6 >> 2 is just * 3 / 2. 

We could do the same with x * 150 / 100.

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:669]: I am in fact working on this in another PR but that is taking more work than expected, so I figure at least do some baby steps so if something has to be reverted, it won't be this PR.

Comment (nikic) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:669]: I agree that doing this in a separate PR makes sense. Just wanted to mention the possibility.

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:669]: The point is, I very much considered this, and not just for shifts, as this transform came about when I was reading code that did 50% increases by doing x = x * 150/100

Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:669]: > This is missing tests for numbers other than 3 and 1. You should test both another combination that works and one that doesn't.

@AtariDreams You can find more cases here: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/607/files

> It's also worth noting that this can still be further generalized. You now handle something like * 3 >> 1, but this also holds for * 6 >> 2 or * 12 >> 3.

We need a canonicalization that converts `(X *nuw/nsw 6) >> 2 -> (X *nuw/nsw 3) >> 1`.

Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:669]: > > This is missing tests for numbers other than 3 and 1. You should test both another combination that works and one that doesn't.
> 
> @AtariDreams You can find more cases here: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/607/files
> 
> > It's also worth noting that this can still be further generalized. You now handle something like * 3 >> 1, but this also holds for * 6 >> 2 or * 12 >> 3.
> 
> We need a canonicalization that converts `(X *nuw/nsw 6) >> 2 -> (X *nuw/nsw 3) >> 1`.

I mean, this is just glorified fraction reduction, which I did in fact start writing, but have not been able to get it to work.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (nikic) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1468]: I'm not entirely sure, but it's possible that the nuw flag makes this safe (by converting problematic undef cases to poison).

File: llvm/test/Transforms/InstCombine/ashr-lshr.ll
Comment (AtariDreams) [reply llvm/test/Transforms/InstCombine/ashr-lshr.ll:669]: > > > This is missing tests for numbers other than 3 and 1. You should test both another combination that works and one that doesn't.
> > 
> > 
> > @AtariDreams You can find more cases here: https://github.com/dtcxzyw/llvm-opt-benchmark/pull/607/files
> > > It's also worth noting that this can still be further generalized. You now handle something like * 3 >> 1, but this also holds for * 6 >> 2 or * 12 >> 3.
> > 
> > 
> > We need a canonicalization that converts `(X *nuw/nsw 6) >> 2 -> (X *nuw/nsw 3) >> 1`.
> 
> I mean, this is just glorified fraction reduction, which I did in fact start writing, but have not been able to get it to work.

So, I am doing that in multiple steps.

**Comment by AtariDreams** [comment]
New tests added!

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1474]: `NewAdd->setHasNoSignedWrap(cast<OverflowingBinaryOperator>(Op0)->hasNoSignedWrap());`

?

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1468]: cc @regehr @nunoplopes 

Can you guys have a look? Thank you!

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1474]: Honestly, I think this looks cleaner, but you do have a point. I should keep the way it was written before.

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1468]: > I'm not entirely sure, but it's possible that the nuw flag makes this safe (by converting problematic undef cases to poison).

It is confusing to me. Is this behavior documented in LangRef?
See https://alive2.llvm.org/ce/z/i2eXpk
```
; bad
define i8 @src1(i8 %x) {
  %shl = shl i8 %x, 1
  ret i8 %shl
}

define i8 @tgt1(i8 %x) {
  %add = add i8 %x, %x
  ret i8 %add
}

; good
define i8 @src2(i8 %x) {
  %shl = shl nuw i8 %x, 1
  ret i8 %shl
}

define i8 @tgt2(i8 %x) {
  %add = add i8 %x, %x
  ret i8 %add
}
```

Comment (regehr) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1468]: Nuno will have to enlighten us, I don't see why `nuw` would stop multiple uses of undef from being a problem, that doesn't make any sense to me

Comment (nunoplopes) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1468]: `shl nuw %x, 1` implies that %x is not undef. This is because if %x is undef, we can pick a value that makes the shift yield poison.
Then we can refine poison with `add %x, %x`.

TL;DR:  `shl nuw undef, 1 -> poison -> add %x, %x`

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1468]: > `shl nuw %x, 1` implies that %x is not undef. This is because if %x is undef, we can pick a value that makes the shift yield poison. Then we can refine poison with `add %x, %x`.
> 
> TL;DR: `shl nuw undef, 1 -> poison -> add %x, %x`

It makes sense to me. Thank you!

**Comment by AtariDreams** [comment]
@dtcxzyw I do not have merge permissions! But I think this is now ready!

**Comment by AtariDreams** [comment]
Is this good?

**Comment by AtariDreams** [comment]
@nikic can we merge this?

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1506]: At least imo, the code would be simpler as:
```
if (match(Op0, m_NUWMul(m_Value(X), m_APInt(MulC))) && BitWidth > 2 &&
    (*MulC - 1).isPowerOf2() && MulC->logBase2() == ShAmtC) {
  auto *BO0 = cast<OverflowingBinaryOperator>(Op0);
  // Look for a "splat" mul pattern - it replicates bits across each half
  // of a value, so a right shift is just a mask of the low bits:
  // lshr i[2N] (mul nuw X, (2^N)+1), N --> and iN X, (2^N)-1
  if (ShAmtC * 2 == BitWidth && BO0->hasNoUnsignedWrap())
    return BinaryOperator::CreateAnd(X, ConstantInt::get(Ty, *MulC - 2));

  // lshr (mul nuw (X, 2^N + 1)), N -> add nuw (X, lshr(X, N))
  if (Op0->hasOneUse() &&
      (BO0->hasNoUnsignedWrap() || &&BO0->hasNoSignedWrap())) {
    auto *NewAdd = BinaryOperator::CreateAdd(
        X,
        Builder.CreateLShr(X, ConstantInt::get(Ty, ShAmtC), "", I.isExact()));
    NewAdd->setHasNoSignedWrap(BO0->hasNoSignedWrap());
    NewAdd->setHasNoUnsignedWrap(BO0->hasNoSignedWrap());
    return NewAdd;
  }
}
```

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1505]: Did an investigation and yes this is true but there are cases where it doesn't have to be exact either and I think that's best for a follow up PR

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1506]: Yeah, but I want to keep it the way it is now in case other new folds come up, so it is easier to extend.

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1506]: In general simpler code w/ less duplication will be easier to extend.

Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1506]: > if (match(Op0, m_NUWMul(m_Value(X), m_APInt(MulC))) && BitWidth > 2 &&
>     (*MulC - 1).isPowerOf2() && MulC->logBase2() == ShAmtC) {
>   auto *BO0 = cast<OverflowingBinaryOperator>(Op0);
>   // Look for a "splat" mul pattern - it replicates bits across each half
>   // of a value, so a right shift is just a mask of the low bits:
>   // lshr i[2N] (mul nuw X, (2^N)+1), N --> and iN X, (2^N)-1
>   if (ShAmtC * 2 == BitWidth && BO0->hasNoUnsignedWrap())
>     return BinaryOperator::CreateAnd(X, ConstantInt::get(Ty, *MulC - 2));
> 
>   // lshr (mul nuw (X, 2^N + 1)), N -> add nuw (X, lshr(X, N))
>   if (Op0->hasOneUse() &&
>       (BO0->hasNoUnsignedWrap() || &&BO0->hasNoSignedWrap())) {
>     auto *NewAdd = BinaryOperator::CreateAdd(
>         X,
>         Builder.CreateLShr(X, ConstantInt::get(Ty, ShAmtC), "", I.isExact()));
>     NewAdd->setHasNoSignedWrap(BO0->hasNoSignedWrap());
>     NewAdd->setHasNoUnsignedWrap(BO0->hasNoSignedWrap());
>     return NewAdd;
>   }
> }

this eliminates the MulC >> ShiftC optimization

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (AtariDreams) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1505]: Fixed!

**Comment by AtariDreams** [comment]
@topperc @dtcxzyw @nikic I addressed all your comments and concerns. Can we please merge this?

File: llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp
Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineShifts.cpp:1506]: That wasn't exact code, just the idea of putting all the `m_Mul` and `(*MulC - 1).isPowerOf2() && MulC->logBase2() == ShAmtC` together to simplify the code.

**Comment by AtariDreams** [comment]
Putting this in hold for a tiny bit as I do one last thing.

**Comment by AtariDreams** [comment]
@nikic I moved the fold to X special cases to analysis, and the rest remain intact. This now is good to merge!

**Comment by AtariDreams** [comment]
@dtcxzyw What do you think about this?

**Comment by AtariDreams** [comment]
@dtcxzyw Can we please benchmark this one more time?

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (topperc) [inline llvm/lib/Analysis/InstructionSimplify.cpp:1495]: affects -> effects

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (topperc) [inline llvm/lib/Analysis/InstructionSimplify.cpp:1484]: This code doesn't create an `and`

File: llvm/lib/Analysis/InstructionSimplify.cpp
Comment (topperc) [inline llvm/lib/Analysis/InstructionSimplify.cpp:1502]: The computeKNownBits part of this doesn't appear to be tested. I cloned your branch and commented this out and nothing failed.

I would think calling computeKnownBits on the lshr would return 0 for this case all by itself.

Comment (AtariDreams) [reply llvm/lib/Analysis/InstructionSimplify.cpp:1502]: it is folding 

 %xx = and i32 %x, 360
  %m = mul i32 %x, 65537
  %t = ashr i32 %m, 16
  
  to ; CHECK-NEXT:    [[M:%.*]] = mul nuw i32 [[X:%.*]], 65537
; CHECK-NEXT:    [[T:%.*]] = ashr i32 [[M]], 16
; CHECK-NEXT:    ret i32 [[T]]

instead of just folding to the and

Comment (topperc) [reply llvm/lib/Analysis/InstructionSimplify.cpp:1502]: This is in `simplifyLShrInst` so why are you showing an `ashr`?

Comment (AtariDreams) [reply llvm/lib/Analysis/InstructionSimplify.cpp:1502]: Fixed it!

Comment (topperc) [reply llvm/lib/Analysis/InstructionSimplify.cpp:1502]: Nevermind. Forget what I said.

**Review by nikic — CHANGES_REQUESTED** [review]
This PR looked sensible the last time I checked it -- now it contains another new random InstSimplify fold and also seems to change some one-use conditions.

Please learn some patience. Don't keep making random changes to your PR just because nobody is around to hit the merge button right this second. Do not ping PRs more often than once a week.

**Comment by nikic** [comment]
You are modifying an existing fold to produce a different result (in a different pass now no less) and adding a new fold in the same PR. That's not going to fly. Please submit two PRs for the two changes.

**Pull request closed without merging**
